### PR TITLE
March-2022-maintenance

### PR DESCRIPTION
**Pull request by marlonbrs**
TLDR: There are some new features that might affect how we currently use strange router. Some features will involve bigger changes than the others but seems doable. The biggest thing in my opinion is we might want to decide on is how we’ll treat childRoutes.

`Index Routes` - this feature might change how we currently handle a child route that matches the parent route’s url. Doesn't seem like it'll involve too much change in strange router's implementation though 
- https://reactrouter.com/docs/en/v6/getting-started/concepts#index-routes
		
 `Layout Routes` - since we’re currently using layout components in strange router, this might be useful but not necessary. Whether we’re using a Layout Route ( route that has no path property and shared across all child routes ) or a normal route with child routes, the structure remains the same where the components that share this Layout will be nested inside it. 

- https://reactrouter.com/docs/en/v6/getting-started/concepts#layout-routes

`Switch component replaced by Routes` - after doing some tests with regards to Nested Routes vs Descendant routes, it seems like the main purpose of descendant routes is for organizing routes. It’s basically a Layout Route that has child routes. The reason I say it’s mainly for organization is that instead of declaring the nested route in top-level, we’re declaring a new route configuration in the component that serves as the parent component to these child routes. I felt it’s necessary to understand these concepts as it changes how we process child routes in place of Switch. 
    
For nested routes  
```
      <Route path=‘/‘ element={<HomePage/>}>    <Route index element={<DefaultPage/>}/>
     <Route path=‘settings’ element={<Settings/>}/>
   </Route> 
```
   
 For descendant routes 
 ```
  // ‘*’ at the end of the path indicates this route has descendant routes
  // ‘*’ can be seen as the replacement of the now deprecated exact prop 
    
  // In App.js:
   <Route path=‘/*’ element={<HomePage/>}/> 
            
   // In HomePage.js:              
   Function HomePage() {                 
      return (
         <>
           <Routes>
               <Route index element={<DefaultPage/>}/>
               <Route path=‘settings’ element={<Settings/>}/>
            </Routes>
  
            <Outlet/>
         </>
      ) 
   }

```
- Nested routes - https://ui.dev/react-router-nested-routes
- Descendant Routes - https://www.robinwieruch.de/react-router-descendant-routes/)

`Outlet component` - the way I see it, this is the component that renders the child route’s component. In strange router, this can be seen as the `children` prop we use in the parent route where child routes get rendered. In terms of implementation, this might have us wrap the component passed in strange router in a higher order component that returns a new component with an Outlet component

- https://reactrouter.com/docs/en/v6/getting-started/concepts#outlets  
     
`Catch All route` - both in nested and descendant routes, catch all route are added as a sibling route to the child routes. The only change I see here is instead of using a redirect component, we’ll render a Route component with a path of `*`

### CODE REVIEW

**Review by zemccartney — COMMENTED** [review]
Yo! I think to bring this home we'll want to go through the migration guide

I expect we'll need a handful of changes in both `lib/index.js` and `demo/index.html` to account for whatever breaking changes we see there.

I haven't gone through that guide in super-depth, but on a quick skim, a couple of items stand out to me, that I know affect us:

- https://reactrouter.com/docs/en/v6/upgrading/v5#remove-redirects-inside-switch (since we use this pattern)
- https://reactrouter.com/docs/en/v6/upgrading/v5#upgrade-all-switch-elements-to-routes (builds on the previous point, getting rid of `Switch` entirely)

Though I bet there could be more. Lemme know if any questions! Totally down to pair on this anytime today / tomorrow if that'd be helpful 🙏
